### PR TITLE
Simplify boolean return logic

### PR DESCRIPTION
**Pull request by nickdrozd**
I was poking around in the code and I noticed that there are a lot of bare boolean returns. For example:

```python
        if self.subtype_context.keep_erased_types:
            return False
        return True
```

This can be simplified to:

```python
        return not self.subtype_context.keep_erased_types
```

That's a pretty simple case. Other cases are more complicated, but similar in spirit.

This is a refactoring PR that should not change any behavior. By my count it results in a net reduction of 32 `return` statements.

If this sort of change is welcome, I can keep going.

### CODE REVIEW

File: mypy/subtypes.py [hunk 1]
```diff
@@ -403,48 +403,43 @@ def visit_any(self, left: AnyType) -> bool:
...
+        )
 
     def visit_erased_type(self, left: ErasedType) -> bool:
         # This may be encountered during type inference. The result probably doesn't
         # matter much.
         # TODO: it actually does matter, figure out more principled logic about this.
-        if self.subtype_context.keep_erased_types:
-            return False
-        return True
+        return not self.subtype_context.keep_erased_types
```
Comment (hauntsaninja) [inline mypy/subtypes.py:431]: keep

File: mypy/subtypes.py [hunk 2]
```diff
@@ -403,48 +403,43 @@ def visit_any(self, left: AnyType) -> bool:
...
-                # can by a subtype of it, even with an `Any` fallback.
-                # This special case is needed to treat descriptors in classes with
-                # dynamic base classes correctly, see #5456.
-                return False
-            return True
+            # NOTE: `None` is a *non-subclassable* singleton, therefore no class
+            # can by a subtype of it, even with an `Any` fallback.
+            # This special case is needed to treat descriptors in classes with
+            # dynamic base classes correctly, see #5456.
+            return not isinstance(self.right, NoneType)
```
Comment (hauntsaninja) [inline mypy/subtypes.py:442]: keep

File: mypy/subtypes.py [hunk 3]
```diff
@@ -513,11 +508,7 @@ def check_mixed(
                             isinstance(unpacked_type, Instance)
                             and unpacked_type.type.fullname == "builtins.tuple"
                         ):
-                            if not all(
-                                is_equivalent(l, unpacked_type.args[0]) for l in compare_to
-                            ):
-                                return False
-                            return True
+                            return all(is_equivalent(l, unpacked_type.args[0]) for l in compare_to)
```
Comment (hauntsaninja) [inline mypy/subtypes.py:511]: keep

File: mypy/subtypes.py [hunk 4]
```diff
@@ -735,15 +716,12 @@ def visit_tuple_type(self, left: TupleType) -> bool:
...
-                return True
-            return False
+            return self._is_subtype(mypy.typeops.tuple_fallback(left), right)
         elif isinstance(right, TupleType):
             if len(left.items) != len(right.items):
                 return False
-            for l, r in zip(left.items, right.items):
-                if not self._is_subtype(l, r):
-                    return False
+            if any(not self._is_subtype(l, r) for l, r in zip(left.items, right.items)):
```
Comment (hauntsaninja) [inline mypy/subtypes.py:723]: keep

File: mypy/subtypes.py [hunk 5]
```diff
@@ -752,9 +730,7 @@ def visit_tuple_type(self, left: TupleType) -> bool:
                 # join(Union[int, C], Union[str, C]) == Union[int, str, C].
                 return True
             lfallback = mypy.typeops.tuple_fallback(left)
-            if not self._is_subtype(lfallback, rfallback):
-                return False
-            return True
+            return self._is_subtype(lfallback, rfallback)
```
Comment (hauntsaninja) [inline mypy/subtypes.py:733]: keep

File: mypy/subtypes.py [hunk 6]
```diff
@@ -811,10 +787,7 @@ def visit_overloaded(self, left: Overloaded) -> bool:
                     return True
             return self._is_subtype(left.fallback, right)
         elif isinstance(right, CallableType):
-            for item in left.items:
-                if self._is_subtype(item, right):
-                    return True
-            return False
+            return any(self._is_subtype(item, right) for item in left.items)
```
Comment (hauntsaninja) [inline mypy/subtypes.py:790]: keep

File: mypy/subtypes.py [hunk 7]
```diff
@@ -988,9 +961,8 @@ def f(self) -> A: ...
         if not members_right.issubset(members_left):
             return False
     assuming = right.type.assuming_proper if proper_subtype else right.type.assuming
-    for (l, r) in reversed(assuming):
-        if l == left and r == right:
-            return True
+    if any(l == left and r == right for (l, r) in reversed(assuming)):
```
Comment (hauntsaninja) [inline mypy/subtypes.py:964]: keep

File: mypy/subtypes.py [hunk 8]
```diff
@@ -1376,8 +1348,7 @@ def g(x: int) -> int: ...
         unified = unify_generic_callable(left, right, ignore_return=ignore_return)
         if unified is None:
             return False
-        else:
-            left = unified
+        left = unified
```
Comment (hauntsaninja) [inline mypy/subtypes.py:1351]: keep

File: mypy/types.py [hunk 9]
```diff
@@ -437,14 +437,12 @@ def __repr__(self) -> str:
...
     def __eq__(self, other: object) -> bool:
-        if isinstance(other, TypeVarId):
-            return (
-                self.raw_id == other.raw_id
-                and self.meta_level == other.meta_level
-                and self.namespace == other.namespace
-            )
-        else:
-            return False
+        return (
```
Comment (hauntsaninja) [inline mypy/types.py:440]: keep

File: mypy/types.py [hunk 10]
```diff
@@ -910,9 +908,7 @@ def __hash__(self) -> int:
         return hash(tuple(self.items))
 
     def __eq__(self, other: object) -> bool:
-        if not isinstance(other, TypeList):
-            return False
-        return self.items == other.items
+        return isinstance(other, TypeList) and self.items == other.items
```
Comment (hauntsaninja) [inline mypy/types.py:911]: keep

File: mypy/types.py [hunk 11]
```diff
@@ -2263,16 +2261,19 @@ def __hash__(self) -> int:
...
     def __eq__(self, other: object) -> bool:
-        if isinstance(other, TypedDictType):
-            if frozenset(self.items.keys()) != frozenset(other.items.keys()):
-                return False
-            for (_, left_item_type, right_item_type) in self.zip(other):
-                if not left_item_type == right_item_type:
-                    return False
-            return self.fallback == other.fallback and self.required_keys == other.required_keys
-        else:
+        if not isinstance(other, TypedDictType):
```
Comment (hauntsaninja) [inline mypy/types.py:2264]: keep

File: mypy/types.py [hunk 12]
```diff
@@ -3346,21 +3344,19 @@ def is_literal_type(typ: ProperType, fallback_fullname: str, value: LiteralValue
     """Check if this type is a LiteralType with the given fallback type and value."""
     if isinstance(typ, Instance) and typ.last_known_value:
         typ = typ.last_known_value
-    if not isinstance(typ, LiteralType):
-        return False
-    if typ.fallback.type.fullname != fallback_fullname:
-        return False
-    return typ.value == value
+    return (
```
Comment (hauntsaninja) [inline mypy/types.py:3347]: keep

**Review by hauntsaninja — CHANGES_REQUESTED** [review]
Thanks, but several of these make the code less readable. In general, some of these sequences of if conditions are pretty subtle and it's much easier to read them one at a time than to have them compressed into a single statement. I've commented on the ones I think are worthwhile, if you want to pursue this PR, please remove the others :-)

**Comment by nickdrozd** [comment]
Okay, I think I got all of them.

File: mypy/types.py [hunk 13]
```diff
@@ -2263,16 +2259,19 @@ def __hash__(self) -> int:
...
-                    return False
-            return self.fallback == other.fallback and self.required_keys == other.required_keys
-        else:
+        if not isinstance(other, TypedDictType):
             return NotImplemented
 
+        return (
+            frozenset(self.items.keys()) == frozenset(other.items.keys())
+            and all(
+                not left_item_type == right_item_type
```
Comment (hauntsaninja) [inline mypy/types.py:2268]: This needs fixing :-)

Comment (nickdrozd) [reply mypy/types.py]: Fixed (I hope)

**Comment by nickdrozd** [comment]
I fixed the existing changes and added a few more. All the new changes are pretty small, and they can be kept or not.

In the spirit of simplified boolean logic, it would be easier if you mark the ones to reject, rather than the ones to keep :smile:

File: mypy/checker.py [hunk 14]
```diff
@@ -6152,13 +6147,12 @@ def handle_partial_var_type(
...
-                if base.get(var.name) is not None:
-                    return True
-            if var.info.fallback_to_any:
-                return True
-        return False
+        if not var.info:
+            return False
+        return (
+            any(base.get(var.name) is not None for base in var.info.mro[1:])
+            or var.info.fallback_to_any
```
Comment (hauntsaninja) [inline mypy/checker.py:6154]:
```suggestion
            var.info.fallback_to_any
            or any(base.get(var.name) is not None for base in var.info.mro[1:])
```

File: mypy/meet.py [hunk 15]
```diff
@@ -535,9 +535,7 @@ def are_typed_dicts_overlapping(
     """Returns 'true' if left and right are overlapping TypeDictTypes."""
     # All required keys in left are present and overlapping with something in right
     for key in left.required_keys:
-        if key not in right.items:
-            return False
-        if not is_overlapping_types(
+        if key not in right.items or not is_overlapping_types(
```
Comment (hauntsaninja) [inline mypy/meet.py:538]: revert

File: mypy/meet.py [hunk 16]
```diff
@@ -547,9 +545,7 @@ def are_typed_dicts_overlapping(
 
     # Repeat check in the other direction
     for key in right.required_keys:
-        if key not in left.items:
-            return False
-        if not is_overlapping_types(
+        if key not in left.items or not is_overlapping_types(
```
Comment (hauntsaninja) [inline mypy/meet.py:548]: revert

File: mypy/checker.py [hunk 17]
```diff
@@ -6348,14 +6342,11 @@ def conditional_types_with_intersection(
     def is_writable_attribute(self, node: Node) -> bool:
         """Check if an attribute is writable"""
         if isinstance(node, Var):
-            if node.is_property and not node.is_settable_property:
-                return False
-            return True
+            return not node.is_property or node.is_settable_property
```
Comment (hauntsaninja) [inline mypy/checker.py:6345]: revert

File: mypy/stubgen.py [hunk 18]
```diff
@@ -1489,32 +1486,28 @@ def find_module_paths_using_imports(
...
-            ".testing",
-            "_tests",
-            "_test_suite",
-            "test_util",
-            "test_utils",
-            "test_base",
-            ".__main__",
-            ".conftest",  # Used by pytest
-            ".setup",  # Typically an install script
+    return (
```
Comment (hauntsaninja) [inline mypy/stubgen.py:1489]: revert

File: mypyc/ir/class_ir.py [hunk 19]
```diff
@@ -248,15 +248,12 @@ def is_method_final(self, name: str) -> bool:
...
             return False
 
-        if self.has_method(name):
-            method_decl = self.method_decl(name)
-            for subc in subs:
-                if subc.method_decl(name) != method_decl:
-                    return False
-            return True
-        else:
+        if not self.has_method(name):
```
Comment (hauntsaninja) [inline mypyc/ir/class_ir.py:251]: revert

File: mypyc/ir/class_ir.py [hunk 20]
```diff
@@ -265,15 +262,10 @@ def has_attr(self, name: str) -> bool:
...
-            if name in ir.deletable:
-                return True
-        return False
+        return any(name in ir.deletable for ir in self.mro)
 
     def is_always_defined(self, name: str) -> bool:
-        if self.is_deletable(name):
-            return False
-        return name in self._always_initialized_attrs
+        return not self.is_deletable(name) and name in self._always_initialized_attrs
```
Comment (hauntsaninja) [inline mypyc/ir/class_ir.py:268]: revert

File: mypyc/irbuild/ast_helpers.py [hunk 21]
```diff
@@ -115,6 +115,4 @@ def is_borrow_friendly_expr(self: IRBuilder, expr: Expression) -> bool:
         if isinstance(expr.node, Var) and expr.kind == LDEF:
             # Local variable reference can be borrowed
             return True
-    if isinstance(expr, MemberExpr) and self.is_native_attr_ref(expr):
-        return True
-    return False
+    return isinstance(expr, MemberExpr) and self.is_native_attr_ref(expr)
```
Comment (hauntsaninja) [inline mypyc/irbuild/ast_helpers.py:118]: revert

**Review by hauntsaninja — COMMENTED** [review]
Thanks, reviewed the new changes. Please don't add any others in this PR :-)

File: mypy/checker.py (hunk 14 above)
Comment (nickdrozd) [reply mypy/checker.py]: This seems like a reasonable change, although unlike the other ones it isn't strictly equivalent.

**Review by hauntsaninja — APPROVED** [review]
Thank you!

**Comment by nickdrozd** [comment]
Thanks for reviewing!

**Pull request merged**